README.md
[![All Contributors](https://img.shields.io/badge/all_contributors-1-orange.svg?style=flat-square)](#contributors)
[![Build Status](https://travis-ci.org/tagraha/react-start.svg?branch=master)](https://travis-ci.org/tagraha/react-start)
[![CircleCI](https://circleci.com/gh/tagraha/react-start.svg?style=svg)](https://circleci.com/gh/tagraha/react-start)
[![style: styled-components](https://img.shields.io/badge/style-%F0%9F%92%85%20styled--components-orange.svg?colorB=daa357&colorA=db748e)](https://github.com/styled-components/styled-components)
[![Known Vulnerabilities](https://snyk.io//test/github/tagraha/react-start/badge.svg?targetFile=package.json)](https://snyk.io//test/github/tagraha/react-start?targetFile=package.json)

<a 
target="_blank"
href="https://david-dm.org/tagraha/react-start">
  <img src="https://david-dm.org/tagraha/react-start.svg" alt="Dependency Status" />
</a>
<a 
target="_blank"
href="https://david-dm.org/tagraha/react-start?type=dev">
  <img src="https://david-dm.org/tagraha/react-start/dev-status.svg" alt="Dev Dependency Status" />
</a>

Heavily modification of react-universaly
----------------------------------------
React boilerplate with code spliting, asynchronous data fetching (universal/isomorphic rendering)

Heads up
--------
We are using react ^16.3.0. You might want to read this ([React v16.3.0: New lifecycles and context API](https://reactjs.org/blog/2018/03/29/react-v-16-3.html)) before starting playing around with this boilerplate.

Installation
------------
```bash
git clone git@github.com:tagraha/react-start.git
cd react-start
npm install
```

at this point make `.env` file on your root directory. you can copy it from `.env_example` file.

```bash
npm run develop
```

open up your browser and navigate to `localhost:1337`

Production script command
-------------------------
```bash
npm run build
npm start
```

-------------

The App
-------
Navigate to `shared/container/DemoApp` and start exploring the code

Configuration
-------------
you can configure like disabling SSR mode, renaming service worker file etc with `value.js` file. here's the the quick look of the file

```javascript

const values = {
  clientConfigFilter: {
    // This is here as an example showing that you can expose variables
    // that were potentially provivded by the environment
    welcomeMessage: true,
    // We only need to expose the enabled flag of the service worker.
    serviceWorker: {
      enabled: true,
    },
    // We need to expose all the polyfill.io settings.
    polyfillIO: true,
    // We need to expose all the htmlPage settings.
    htmlPage: true,
  },

  // The host on which the server should run.
  host: EnvVars.string('HOST', 'localhost'),

  // The port on which the server should run.
  port: EnvVars.number('PORT', 1337),

  // The port on which the client bundle development server should run.
  clientDevServerPort: EnvVars.number('CLIENT_DEV_PORT', 7331),
  welcomeMessage: EnvVars.string('WELCOME_MSG', 'Hello world!'),

  // Disable server side rendering?
  disableSSR: false,
  browserCacheMaxAge: '365d',
}
```

TODOS
-----
- [x] attach redux
- [x] make it async data fetch
- [x] upgrade to webpack 4 (perks)
- [x] unit testing
- [ ] flowjs type checker
- [ ] redux-saga over redux-thunk?

Flow Coverage
-------------
we're currently working on flow coverage. you can switch to branch `feature/flow` to take a look at the code
![flow-coverage](https://raw.githubusercontent.com/tagraha/react-start/master/flow-coverage.png)


## Contributors

Thanks goes to these wonderful people ([emoji key](https://github.com/kentcdodds/all-contributors#emoji-key)):

<!-- ALL-CONTRIBUTORS-LIST:START - Do not remove or modify this section -->
<!-- prettier-ignore -->
| [<img src="https://avatars0.githubusercontent.com/u/3034375?v=4" width="100px;"/><br /><sub><b>Tirta Nugraha</b></sub>](http://www.betotally.com/)<br />[💻](https://github.com/tagraha/react-start/commits?author=tagraha "Code") [📖](https://github.com/tagraha/react-start/commits?author=tagraha "Documentation") [🔌](#plugin-tagraha "Plugin/utility libraries") [👀](#review-tagraha "Reviewed Pull Requests") [⚠️](https://github.com/tagraha/react-start/commits?author=tagraha "Tests") |
| :---: |
<!-- ALL-CONTRIBUTORS-LIST:END -->

This project follows the [all-contributors](https://github.com/kentcdodds/all-contributors) specification. Contributions of any kind welcome!


CHANGELOG.md
### 2.9.0 (2020-03-30)

##### Bug Fixes

* **package:**  fix vulnerable warning from package ([31bad109](https://github.com/tagraha/react-start/commit/31bad109947075b0b8b652bd2bce4ff85590de2f))

##### Other Changes

* **package:**
  *  update webpack bundle analyzer ([e85c063d](https://github.com/tagraha/react-start/commit/e85c063d42da9219420a25d804e891afeeea4b22))
  *  update react-dev-utils ([3b9a6e94](https://github.com/tagraha/react-start/commit/3b9a6e9443cb17f922ac718ef85effdd18321f00))
  *  update style-loader ([c62b8e93](https://github.com/tagraha/react-start/commit/c62b8e936b75d2ef3dde547dbcf74645c60d32cf))
  *  update eslint plugin ([f91d2ede](https://github.com/tagraha/react-start/commit/f91d2edea337c99b41839396e6216f707fd2ccde))
  *  update html webpack plugin ([3d8ff729](https://github.com/tagraha/react-start/commit/3d8ff729fbb43979dcea396f0682d3af177f4224))
  *  update prettier ([3e3f49d0](https://github.com/tagraha/react-start/commit/3e3f49d03535baf1952b541792510e885e78e425))
  *  update webpack ([e2b900cb](https://github.com/tagraha/react-start/commit/e2b900cb1d6a2b045b924dc036f560bf0a940f42))
  *  update helmet ([c05cec2b](https://github.com/tagraha/react-start/commit/c05cec2b2d75bac01a76045d0182bbae37f55eab))
  *  update react ([d61c32ab](https://github.com/tagraha/react-start/commit/d61c32ab68f661db0e7e29bdaf926fa32390fc59))
  *  update terser-webpack-plugin ([5495618a](https://github.com/tagraha/react-start/commit/5495618aee2177c358f4ed9667fe3090294b573f))
  *  update react & webpack ([a5ed90ab](https://github.com/tagraha/react-start/commit/a5ed90abaefc40f6cd82a2e616939c3bc04fc9bd))
  *  update axios ([a2d201ba](https://github.com/tagraha/react-start/commit/a2d201ba3335c00a975440de89f0557511c5f264))

##### Refactors

* **hook:**  update class component to hook ([5e0b0ed9](https://github.com/tagraha/react-start/commit/5e0b0ed9615685ddc46db3ea0702fbe6f2de655f))

#### 2.8.2 (2020-01-10)

##### Other Changes

* **package:**
  *  update husky ([d558692f](https://github.com/tagraha/react-start/commit/d558692f15cfef09a51924005a424566af5915be))
  *  update css-loader ([505d3d88](https://github.com/tagraha/react-start/commit/505d3d881733c1cb847c8ab97b1a170e3c562491))
  *  update style-loader ([af961625](https://github.com/tagraha/react-start/commit/af961625280437f47f9d856ce95adae10cd8ba7f))
  *  update terser-webpack-plugin ([212dcb3c](https://github.com/tagraha/react-start/commit/212dcb3c0cbfa30af0df010bdf7851542cc680cc))

#### 2.8.1 (2019-12-31)

##### Other Changes

* **package:**
  *  update redux ([e91fc594](https://github.com/tagraha/react-start/commit/e91fc59409973665ac1de823ed35c0395a4a9b16))
  *  update react-redux ([a8ed3537](https://github.com/tagraha/react-start/commit/a8ed353734bc19fca76e6fb77334f824fa912857))
  *  update source-map-support ([6a284757](https://github.com/tagraha/react-start/commit/6a2847575323be0c7501287463d81d3c8ffe413b))
  *  update css-loader ([09345433](https://github.com/tagraha/react-start/commit/093454332c09b96ea6a9335a70d8a7dac0c9eb9f))
  *  update mini-css-extract-plugin ([19841e9c](https://github.com/tagraha/react-start/commit/19841e9c08e64b9a8e3bb105be44c0c0931e1fda))
  *  update react-dev-utils ([8ce5d0b5](https://github.com/tagraha/react-start/commit/8ce5d0b5edbc87fa47c080f51bd7584176b8b899))
  *  update webpack ([e43342da](https://github.com/tagraha/react-start/commit/e43342da05c4e09b1fea2e033008a2dc08f7d295))
  *  update webpack ([58ab4e2b](https://github.com/tagraha/react-start/commit/58ab4e2b89a54ca76f5d25d3c64e05e5b3180ce6))
  *  update eslint ([86491b9d](https://github.com/tagraha/react-start/commit/86491b9d55a31593fbb25845c86ebfd2472e30c8))
  *  update jest ([cad4500e](https://github.com/tagraha/react-start/commit/cad4500e8f029cf8461c2ae3430f8f72fca19a55))

### 2.8.0 (2019-12-13)

##### New Features

* **babel:**  upgrade to babel 7 ([dc4e3b7c](https://github.com/tagraha/react-start/commit/dc4e3b7c7742a385b40cebd2939f611ae38ed588))

##### Other Changes

* **package:**
  *  update react-hot-loader ([3f6ce347](https://github.com/tagraha/react-start/commit/3f6ce347261c88ccf63f203503e84c769802fc2d))
  *  update serialize-javascript ([4679f32b](https://github.com/tagraha/react-start/commit/4679f32b4b029f3cf0c54781d4152bb7b9a8cc9c))
  *  update style-loader ([053eeb98](https://github.com/tagraha/react-start/commit/053eeb9811292039145051a80dd9b078b1887b6c))
  *  update terser-webpack-plugin ([f6c411d3](https://github.com/tagraha/react-start/commit/f6c411d33a2bd9be4c5eb90f13bb9b0100247fe5))
  *  update webpack-dev-middleware ([e1d1d537](https://github.com/tagraha/react-start/commit/e1d1d5372a3a916965c8566ac113662227ab947a))
  *  update webpack bundle analyzer ([267d29b0](https://github.com/tagraha/react-start/commit/267d29b04d91f40e893515316ee3ec8976b7c4b9))
  *  update helmet ([334fc212](https://github.com/tagraha/react-start/commit/334fc21214ec55f7f1abc3698dc796bc052c4462))
  *  update file-loader ([6d4c184d](https://github.com/tagraha/react-start/commit/6d4c184d013c5615125ce63ef9542a34b9854ab2))
  *  update react-dev-utils ([38621e71](https://github.com/tagraha/react-start/commit/38621e712a4e4bee844734bbd307eca0f8768ce0))
  *  update style-loader ([6ee9a5af](https://github.com/tagraha/react-start/commit/6ee9a5afcf870b1d4eb50da36c1989184bf97041))
  *  update react ([c2cf1a1a](https://github.com/tagraha/react-start/commit/c2cf1a1a6179ae79c9021b49c8d5f8e21f0a3298))
  *  update eslint-plugin-react ([d85b0d4f](https://github.com/tagraha/react-start/commit/d85b0d4f6f35bbf6d4e4c0ae988c11f8a8910e8e))
  *  update terser-webpack-plugin ([d59d5e0b](https://github.com/tagraha/react-start/commit/d59d5e0b1b418d2915ee2bc4cbaa80d1c314f6d5))
  *  update webpack ([5ee5b479](https://github.com/tagraha/react-start/commit/5ee5b479de682db9d1125a46fb4895656f99acf3))
  *  update react, react-dom & react-test-renderer ([23d70049](https://github.com/tagraha/react-start/commit/23d700492ae44c031b56f977867163001f19b18f))

config/values.js
/**
 * Project Configuration.
 *
 * NOTE: All file/folder paths should be relative to the project root. The
 * absolute paths should be resolved during runtime by our build internal/server.
 */

import * as EnvVars from './utils/envVars';

const values = {
  // The configuration values that should be exposed to our client bundle.
  // This value gets passed through the /shared/utils/objects/filterWithRules
  // util to create a filter object that can be serialised and included
  // with our client bundle.
  clientConfigFilter: {
    // This is here as an example showing that you can expose variables
    // that were potentially provivded by the environment
    welcomeMessage: true,
    // We only need to expose the enabled flag of the service worker.
    serviceWorker: {
      enabled: true,
    },
    // We need to expose all the polyfill.io settings.
    polyfillIO: true,
    // We need to expose all the htmlPage settings.
    htmlPage: true,
  },

  // The host on which the server should run.
  host: EnvVars.string('HOST', 'localhost'),

  // The port on which the server should run.
  port: EnvVars.number('PORT', 1337),

  // The port on which the client bundle development server should run.
  clientDevServerPort: EnvVars.number('CLIENT_DEV_PORT', 7331),

  // This is an example environment variable which is used within the react
  // application to demonstrate the usage of environment variables across
  // the client and server bundles.
  welcomeMessage: EnvVars.string('WELCOME_MSG', 'Hello world!'),

  // Disable server side rendering?
  disableSSR: false,

  // How long should we set the browser cache for the served assets?
  // Don't worry, we add hashes to the files, so if they change the new files
  // will be served to browsers.
  // We are using the "ms" format to set the length.
  // @see https://www.npmjs.com/package/ms
  browserCacheMaxAge: '365d',

  // We use the polyfill.io service which provides the polyfills that a
  // client needs, which is far more optimal than the large output
  // generated by babel-polyfill.
  // Note: we have to keep this seperate from our "htmlPage" configuration
  // as the polyfill needs to be loaded BEFORE any of our other javascript
  // gets parsed.
  polyfillIO: {
    enabled: true,
    url: '//cdn.polyfill.io/v2/polyfill.min.js',
    // Reference https://qa.polyfill.io/v2/docs/features for a full list
    // of features.
    features: [
      // The default list.
      'default',
      'es6',
    ],
  },

  // Basic configuration for the HTML page that hosts our application.
  // We make use of react-helmet to consume the values below.
  // @see https://github.com/nfl/react-helmet
  htmlPage: {
    titleTemplate: 'React, start - %s',
    defaultTitle: 'React, start',
    description:
      'A starter kit giving you the minimum requirements for a production ready universal react application.',
  },

  // Content Security Policy (CSP)
  // @see server/middleware/security for more info.
  cspExtensions: {
    childSrc: [],
    connectSrc: [],
    defaultSrc: [],
    fontSrc: ['fonts.googleapis.com/css', 'fonts.gstatic.com'],
    imgSrc: [],
    mediaSrc: [],
    manifestSrc: [],
    objectSrc: [],
    scriptSrc: [
      // Allow scripts from cdn.polyfill.io so that we can import the
      // polyfill.
      'cdn.polyfill.io',
    ],
    styleSrc: [
      'cdn.rawgit.com/milligram/milligram/master/dist/milligram.min.css',
      'fonts.googleapis.com/css',
    ],
  },

  // Path to the public assets that will be served off the root of the
  // HTTP server.
  publicAssetsPath: './public',

  // Where does our build output live?
  buildOutputPath: './build',

  // Do you want to included source maps for optimised builds of the client
  // bundle?
  includeSourceMapsForOptimisedClientBundle: false,

  // These extensions are tried when resolving src files for our bundles..
  bundleSrcTypes: ['js', 'jsx', 'json'],

  // What should we name the json output file that webpack generates
  // containing details of all output files for a bundle?
  bundleAssetsFileName: 'assets.json',

  // node_modules are not included in any bundles that target "node" as a
  // runtime (e.g.. the server bundle) as including them often breaks builds
  // due to thinks like require statements containing expressions..
  // However. some of the modules contain files need to be processed by
  // one of our Webpack loaders (e.g. CSS). Add any file types to the list
  // below to allow them to be processed by Webpack.
  nodeExternalsFileTypeWhitelist: [
    /\.(eot|woff|woff2|ttf|otf)$/,
    /\.(svg|png|jpg|jpeg|gif|ico)$/,
    /\.(mp4|mp3|ogg|swf|webp)$/,
    /\.(css|scss|sass|sss|less)$/,
  ],

  // Note: you can only have a single service worker instance.  Our service
  // worker implementation is bound to the "client" and "server" bundles.
  // It includes the "client" bundle assets, as well as the public folder assets,
  // and it is served by the "server" bundle.
  serviceWorker: {
    // Enabled?
    enabled: true,
    // Service worker name
    fileName: 'sw.js',
    // Paths to the public assets which should be included within our
    // service worker. Relative to our public folder path, and accepts glob
    // syntax.
    includePublicAssets: [
      // NOTE: This will include ALL of our public folder assets.  We do
      // a glob pull of them and then map them to /foo paths as all the
      // public folder assets get served off the root of our application.
      // You may or may not want to be including these assets.  Feel free
      // to remove this or instead include only a very specific set of
      // assets.
      './**/*',
    ],
    // Offline page file name.
    offlinePageFileName: 'offline.html',
  },

  bundles: {
    client: {
      // Src entry file.
      srcEntryFile: './client/index.js',

      // Src paths.
      srcPaths: [
        './client',
        './shared',
        // The service worker offline page generation needs access to the
        // config folder.  Don't worry we have guards within the config files
        // to ensure they never get included in a client bundle.
        './config',
      ],

      // Where does the client bundle output live?
      outputPath: './build/client',

      // What is the public http path at which we must serve the bundle from?
      webPath: '/client/',

      // Configuration settings for the development vendor DLL. This will be created
      // by our development server and provides an improved dev experience
      // by decreasing the number of modules that webpack needs to process
      // for every rebuild of our client bundle.  It by default uses the
      // dependencies configured in package.json however you can customise
      // which of these dependencies are excluded, whilst also being able to
      // specify the inclusion of additional modules below.
      devVendorDLL: {
        // Enabled?
        enabled: true,

        // Specify any dependencies that you would like to include in the
        // Vendor DLL.
        //
        // NOTE: It is also possible that some modules require specific
        // webpack loaders in order to be processed (e.g. CSS/SASS etc).
        // For these cases you don't want to include them in the Vendor DLL.
        include: [
          'react-async-component',
          'react',
          'react-dom',
          'react-helmet',
          'react-router-dom',
        ],

        // The name of the vendor DLL.
        name: '__dev_vendor_dll__',
      },
    },

    server: {
      // Src entry file.
      srcEntryFile: './server/index.js',

      // Src paths.
      srcPaths: ['./server', './shared', './config'],

      // Where does the server bundle output live?
      outputPath: './build/server',
    },
  },

  additionalNodeBundles: {
    // NOTE: The webpack configuration and build scripts have been built so
    // that you can add arbitrary additional node bundle configurations here.
    //
    // A common requirement for larger projects is to add additional "node"
    // target bundles (e.g an APi server endpoint). Therefore flexibility has been
    // baked into our webpack config factory to allow for this.
    //
    // Simply define additional configurations similar to below.  The development
    // server will manage starting them up for you.  The only requirement is that
    // within the entry for each bundle you create and return the "express"
    // listener.
    /*
    apiServer: {
      srcEntryFile: './api/index.js',
      srcPaths: [
        './api',
        './shared',
        './config',
      ],
      outputPath: './build/api',
    }
    */
  },

  // These plugin definitions provide you with advanced hooks into customising
  // the project without having to reach into the internals of the tools.
  //
  // We have decided to create this plugin approach so that you can come to
  // a centralised configuration folder to do most of your application
  // configuration adjustments.  Additionally it helps to make merging
  // from the origin starter kit a bit easier.
  plugins: {
    // This plugin allows you to provide final adjustments your babel
    // configurations for each bundle before they get processed.
    //
    // This function will be called once for each for your bundles.  It will be
    // provided the current webpack config, as well as the buildOptions which
    // detail which bundle and mode is being targetted for the current function run.
    babelConfig: (babelConfig, buildOptions) => {
      // eslint-disable-next-line no-unused-vars
      const { target, mode } = buildOptions;

      // Example
      /*
      if (target === 'server' && mode === 'development') {
        babelConfig.presets.push('foo');
      }
     */

      return babelConfig;
    },

    // This plugin allows you to provide final adjustments your webpack
    // configurations for each bundle before they get processed.
    //
    // I would recommend looking at the "webpack-merge" module to help you with
    // merging modifications to each config.
    //
    // This function will be called once for each for your bundles.  It will be
    // provided the current webpack config, as well as the buildOptions which
    // detail which bundle and mode is being targetted for the current function run.
    webpackConfig: (webpackConfig, buildOptions) => {
      // eslint-disable-next-line no-unused-vars
      const { target, mode } = buildOptions;

      // Example:
      /*
      if (target === 'server' && mode === 'development') {
        webpackConfig.plugins.push(new MyCoolWebpackPlugin());
      }
      */

      // Debugging/Logging Example:
      /*
      if (target === 'server') {
        console.log(JSON.stringify(webpackConfig, null, 4));
      }
      */

      return webpackConfig;
    },
  },
};

// This protects us from accidentally including this configuration in our
// client bundle. That would be a big NO NO to do. :)
if (process.env.BUILD_FLAG_IS_CLIENT === 'true') {
  throw new Error(
    "You shouldn't be importing the `<projectroot>/config/values.js` directly into code that will be included in your 'client' bundle as the configuration object will be sent to user's browsers. This could be a security risk! Instead, use the `config` helper function located at `<projectroot>/config/index.js`.",
  );
}

export default values;
